pkg/controller/signup_test.go

```diff
@@ -212,9 +217,335 @@ func (s *TestSignupSuite) TestSignupGetHandler() {
+		ctx.Params = append(ctx.Params, gin.Param{
+			Key:   "code",
+			Value: "555555",
+		})
+
+		handler(ctx)
+
+		// Check the status code is what we expect.
+		require.Equal(s.T(), http.StatusInternalServerError, rr.Code)
+	})
+
+	s.Run("verifycode returns status error", func() {
+		// Create a mock SignupService
+		svc := &FakeSignupService{
+			MockGetUserSignup: func(userID string) (userSignup *crtapi.UserSignup, e error) {
+				return &crtapi.UserSignup{
+					TypeMeta: v1.TypeMeta{},
+					ObjectMeta: v1.ObjectMeta{
+						Name: userID,
+						Annotations: map[string]string{
+							crtapi.UserSignupUserEmailAnnotationKey:        "jsmith@redhat.com",
+							crtapi.UserVerificationAttemptsAnnotationKey:   "0",
+							crtapi.UserSignupVerificationCodeAnnotationKey: "333333",
+							crtapi.UserVerificationExpiryAnnotationKey:     time.Now().Add(10 * time.Second).Format(verification.TimestampLayout),
+						},
+					},
+					Spec:   crtapi.UserSignupSpec{},
+					Status: crtapi.UserSignupStatus{},
+				}, nil
+			},
+			MockUpdateUserSignup: func(userSignup *crtapi.UserSignup) (userSignup2 *crtapi.UserSignup, e error) {
+				return userSignup, nil
+			},
+		}
+
+		// Create a mock VerificationService
+		verifyService := &FakeVerificationService{
+			MockVerifyCode: func(ctx *gin.Context, signup *crtapi.UserSignup, code string) error {
+				return errors2.NewTooManyRequestsError("too many requests")
+			},
+		}
+
+		// Create Signup controller instance.
+		ctrl := controller.NewSignup(s.Config, svc, verifyService)
+		handler := gin.HandlerFunc(ctrl.VerifyCodeHandler)
+
+		// We create a ResponseRecorder (which satisfies http.ResponseWriter) to record the response.
+		rr := httptest.NewRecorder()
+		ctx, _ := gin.CreateTestContext(rr)
+
+		req, err := http.NewRequest(http.MethodGet, "/api/v1/signup/verification/333333", nil)
+		require.NoError(s.T(), err)
+
+		ctx.Request = req
+		ctx.Set(context.SubKey, userID)
+		ctx.Params = append(ctx.Params, gin.Param{
+			Key:   "code",
+			Value: "333333",
+		})
+
+		handler(ctx)
+
+		// Check the status code is what we expect.
+		require.Equal(s.T(), http.StatusTooManyRequests, rr.Code)
+	})
+
+	s.Run("verifycode returns non status error", func() {
+		// Create a mock SignupService
+		svc := &FakeSignupService{
+			MockGetUserSignup: func(userID string) (userSignup *crtapi.UserSignup, e error) {
+				return &crtapi.UserSignup{
+					TypeMeta: v1.TypeMeta{},
+					ObjectMeta: v1.ObjectMeta{
+						Name: userID,
+						Annotations: map[string]string{
+							crtapi.UserSignupUserEmailAnnotationKey:        "jsmith@redhat.com",
+							crtapi.UserVerificationAttemptsAnnotationKey:   "0",
+							crtapi.UserSignupVerificationCodeAnnotationKey: "222222",
+							crtapi.UserVerificationExpiryAnnotationKey:     time.Now().Add(10 * time.Second).Format(verification.TimestampLayout),
+						},
+					},
+					Spec:   crtapi.UserSignupSpec{},
+					Status: crtapi.UserSignupStatus{},
+				}, nil
+			},
```

Comment:
I wonder if can also improve the code readability by moving the UserSignup creation to a re-usable function too...

Reply:
We're not creating that many UserSignups, also I like the readability of being able to see the entire object we're creating here rather than putting it in a black box function.